apps/api/src/app/workflows-v2/usecases/create-variables-object/create-variables-object.usecase.ts

```diff
@@ -0,0 +1,121 @@
+import { Injectable } from '@nestjs/common';
+import { ControlValuesRepository } from '@novu/dal';
+import { ControlValuesLevelEnum } from '@novu/shared';
+import { Instrument, InstrumentUsecase } from '@novu/application-generic';
+
+import { keysToObject } from '../../util/utils';
+import { buildVariables } from '../../util/build-variables';
+import { CreateVariablesObjectCommand } from './create-variables-object.command';
+import { isStringifiedMailyJSONContent } from '../../../environments-v1/usecases/output-renderers/maily-to-liquid/wrap-maily-in-liquid.command';
+import { MailyAttrsEnum } from '../../../environments-v1/usecases/output-renderers/maily-to-liquid/maily.types';
+
+@Injectable()
+export class CreateVariablesObject {
+  constructor(private readonly controlValuesRepository: ControlValuesRepository) {}
+
+  @InstrumentUsecase()
+  async execute(command: CreateVariablesObjectCommand): Promise<Record<string, unknown>> {
+    const controlValues = await this.getControlValues(command);
+
+    const variables = this.extractAllVariables(controlValues);
+    const arrayVariables = [
+      ...this.extractMailyAttribute(controlValues, MailyAttrsEnum.EACH_KEY),
+      ...this.extractMailyAttribute(controlValues, MailyAttrsEnum.ID, this.extractArrayPath),
+    ];
+    const showIfVariables = this.extractMailyAttribute(controlValues, MailyAttrsEnum.SHOW_IF_KEY);
+
+    return keysToObject(variables, arrayVariables, showIfVariables);
+  }
+
+  private async getControlValues(command: CreateVariablesObjectCommand) {
+    if (command.controlValues) {
+      return [command.controlValues].flat().flatMap((obj) => Object.values(obj));
+    }
+
+    if (command.workflowId) {
+      const controls = await this.controlValuesRepository.find(
+        {
+          _environmentId: command.environmentId,
+          _organizationId: command.organizationId,
+          _workflowId: command.workflowId,
+          level: ControlValuesLevelEnum.STEP_CONTROLS,
+          controls: { $ne: null },
+        },
+        {
+          controls: 1,
+          _id: 0,
+        }
+      );
+
+      return controls
+        .map((item) => item.controls)
+        .flat()
+        .flatMap((obj) => Object.values(obj));
+    }
+
+    return [];
+  }
+
+  /**
+   * Extracts all variables from control values by parsing handlebars syntax {{variable}}.
+   * Removes duplicates from the final result.
+   *
+   * @example
+   * controlValues = [ "John {{name}}", "Address {{address}} {{address}}", "nothing", 123, true ]
+   * returns = [ "name", "address" ]
+   */
+  @Instrument()
+  private extractAllVariables(controlValues: unknown[]): string[] {
+    const variables = controlValues.flatMap((value) => {
+      const templateVariables = buildVariables(undefined, value);
+
+      return templateVariables.validVariables.map((variable) => variable.name);
+    });
+
+    return [...new Set(variables)];
```

Comment:
I was never a big fan of this array, but I'd prefer to have a more extensible array of objects that allow us to add variable types. Is this the right time to do `[{ name: ...} , { name: ...}]` and add stronger type safety?

Reply:
Yes, now it returns just array of strings/variable names which is not ideal, but for now multiple downstream logic depends on this structure, we can improve it later if needed.